src/cread.c

```diff
@@ -19,105 +19,155 @@
  *
  * @author J. Woollen @date 2012-09-15
  */
-void cewind   (nfile      ) f77int *nfile;              { rewind(pb[*nfile]); }
+void cewind( int nfile ) {
+   rewind(pb[nfile]);
+}
 
 /**
- * This subroutine closes a previously opened BUFR file.
+ * Close a previously opened BUFR file.
  *
- * @param[in] nfile -- f77int*: Internal Fortran I/O stream index associated with BUFR file
+ * @param nfile - File ID.
  *
  * @author J. Woollen @date 2012-09-15
  */
-void closfb   (nfile      ) f77int *nfile;              { fclose(pb[*nfile]); }
+void closfb( int nfile ) {
+   fclose(pb[nfile]);
+}
 
 /**
- * This function reads the next message from a BUFR file that was previously opened for reading.
- *
- * @param[in] nfile -- f77int*: Internal Fortran I/O stream index associated with BUFR file
- * @param[out] bufr -- char*: BUFR message
- * @param[in] mxbyt -- f77int*: Dimensioned size (in bytes) of bufr; used by the function to
- *                     ensure that it doesn't overflow the BUFR array
- * @returns crdbufr -- f77int: return code
- *                     - 0 = normal return
- *                     - -1 = end-of-file encountered while reading
- *                     - -2 = I/O error encountered while reading
- *                     - -3 = overflow of bufr array
+ * Read the next message from a BUFR file that was previously opened for reading.
+ *
+ * @param nfile - File ID.
+ * @param bufr - BUFR message
+ * @param mxwrd - Dimensioned size (in integers) of bufr; used by the function to
+ * ensure that it doesn't overflow the array
+ * @returns crdbufr - Return code:
+ * - 0 = normal return
+ * - -1 = end-of-file encountered while reading
+ * - -2 = I/O error encountered while reading
+ * - -3 = overflow of bufr array
  *
  * @author J. Woollen @date 2012-09-15
  */
-f77int crdbufr (nfile,bufr,mxbyt)
-f77int *nfile; f77int *mxbyt; char *bufr;
-{  f77int  nbyt; f77int  nb; f77int wkint[2]; fpos_t nxtpos;
-   fgetpos(pb[*nfile],&lstpos[*nfile]);
-   nb = sizeof(*bufr); bufr[0]=bufr[1];
-   while ( strncmp(bufr,"BUFR",4)!=0)
-   {  memmove(bufr,&bufr[1],3);
-      if(fread(bufr+3,nb,1,pb[*nfile])!=1) return -1;
+int crdbufr( int nfile, int *bufr, int mxwrd ) {
+
+   int nbytrem, nintrem, wkint[2];
+   size_t nb = sizeof(int);
+   char wkchr[17] = "                ";
+   fpos_t nxtpos;
+
+   /* Find the start of the next BUFR message within the file. */
+   fgetpos(pb[nfile],&lstpos[nfile]);
+   while ( strncmp(wkchr,"BUFR",4)!=0) {
+      memmove(wkchr,&wkchr[1],3);
+      if(fread(wkchr+3,1,1,pb[nfile])!=1) return -1;
    }
-   fgetpos(pb[*nfile],&nxtpos); if(fread(bufr+4,nb,4,pb[*nfile])!=4) return -1;
-   memcpy(wkint,bufr,8); nbyt=iupbs01(wkint,"LENM",4)-8;
-   if(nbyt+8>*mxbyt)                           {fsetpos(pb[*nfile],&nxtpos);return -3;};
-   if(fread(bufr+8,nb,nbyt,pb[*nfile])!=nbyt)  {fsetpos(pb[*nfile],&nxtpos);return -2;};
-   if(strncmp(bufr+nbyt+4,"7777",4)!=0)        {fsetpos(pb[*nfile],&nxtpos);return -2;};
```

Comment:
Thanks for breaking this up. It looks like something from the infamous obfuscated C contest...